pkg/cookies/csrf.go

```diff
@@ -0,0 +1,199 @@
+package cookies
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/encryption"
+	"github.com/vmihailenco/msgpack/v4"
+)
+
+var now = time.Now
```

Comment:
Would it make sense to make this part of the csrf itself? With an internal method that allows it to be overridden? How have we done this elsewhere :thinking:

Reply:
I think in a lot of other cases, we frequently passed in a `now time.Time` method argument following the dependency injection pattern that maybe lives on from its Java roots? I think in all cases we've done it, its been for time mocking in tests (this is the same case here, I just use it for cookie new expiration unit tests we hadn't had before).

As an alternative, I've tried this to expose time in tests while keeping the method singatures clean in normal code. For now I've left it private since this only has time manipulated by my `cookies` package unit tests.

After working with `jwt-go` here: https://github.com/oauth2-proxy/mockoidc/blob/65483e3cdabeaeb796f9180b6e4d5dd253d05985/mockoidc.go#L189

I see the huge advantage in potentially exporting `TimeFunc` in our various packages more. Allows us to simplify method signatures (removing the dummy `now time.Now` variables we have bloating them up) while giving access to `time.Now` to manipulate across packages (in case a sub-dependency in a test also needs a custom view of time).

Reply:
@JoelSpeed I think this was the last piece of outstanding feedback in this PR? Did you have any final thoughts on this design?

Reply:
I'd still lean to creating a `timeFunc` member in the `csrf` struct, then rather than overriding a package level var, you override it just for the instance of `csrf`.

I did the following recently when I needed to mock time in my day job, WDYT?

```
type csrf struct {
  nowFunc func() time.Time
}

func (c csrf) now() time.Time {
  if c.nowFunc != nil {
    return c.nowFunc()
  }
  return time.Now()
}
```

Reply:
Yeah - I think this can work too.

I just want us to get away from method signatures that have `now time.Time` as a parameter for the sole purpose of time mocking in tests -- but having the baggage of extra arguments in normal code flows is annoying.

The downside to this & my original code (private package `var now = time.Now`) is we can't fully mock time in units tests that might have dependencies across packages. We definitely have that in our test suite, and in that scenario - I see the testing elegance in public `TimeFunc` variables (like in the `jwt-go` sample I mentioned).

I'll jump to this pattern for now, but let's chat more about a good architecture to standardize how we handle `time.Now` in the codebase.

Reply:
e7d86ce